compiler/rustc_ty_utils/src/representability.rs

```diff
@@ -76,12 +108,129 @@ fn are_inner_types_recursive<'tcx>(
                     Some(hir::Node::Field(field)) => field.ty.span,
                     _ => sp,
                 };
-                match is_type_structurally_recursive(tcx, span, seen, representable_cache, ty) {
-                    Representability::SelfRecursive(_) => {
-                        Representability::SelfRecursive(vec![span])
+
+                let mut result = None;
+
+                // First, we check whether the field type per se is representable.
+                // This catches cases as in #74224 and #84611. There is a special
+                // case related to mutual recursion, though; consider this example:
+                //
+                //   struct A<T> {
+                //       z: T,
+                //       x: B<T>,
+                //   }
+                //
+                //   struct B<T> {
+                //       y: A<T>
+                //   }
+                //
+                // Here, without the following special case, both A and B are
+                // ContainsRecursive, which is a problem because we only report
+                // errors for SelfRecursive. We fix this by detecting this special
+                // case (shadow_seen.first() is the type we are originally
+                // interested in, and if we ever encounter the same AdtDef again,
+                // we know that it must be SelfRecursive) and "forcibly" returning
+                // SelfRecursive (by setting f_res, which tells the calling
+                // invocations of are_inner_types_representable to forward the
+                // result without adjusting).
+                if shadow_seen.len() > 1 && shadow_seen.len() > seen.len() {
+                    if shadow_seen.first().unwrap() == def {
```

Comment:
Sorry, I might not have been clear - I meant that the line above this would not be necessary if the condition was changed as I suggested.

Reply:
I think we can drop the `shadow_seen.len() > 1` part, because `seen` always contains at least one element, so `shadow_seen.len() > seen.len()` implies the former condition.

We can't drop the latter condition, though, because it would lead to nonsensical errors:
```
error[E0072]: recursive type `Foo` has infinite size
 --> t1.rs:1:1
  |
1 | struct Foo { a: u32 }
  | ^^^^^^^^^^      --- recursive without indirection
  | |
  | recursive type has infinite size
```
The problem is that when `are_inner_types_recursive()` is first called, `shadow_seen` equals `seen` and `ty` (and therefore `def`) is the original type (`Foo` in my example), i.e. `shadow_seen.first() == Some(def)` is always true.

The case that we are actually interested in is mutual recursion, where we have recursed to check some field's type (and therefore `shadow_seen.len() > seen.len()`) and _then_ found the original type again, e.g. when checking `B` below, we recurse into `A` and find `B` again:
```rust
struct A<T> {
    z: T,
    x: B<T>,
}

struct B<T> {
    y: A<T>
}
```

Reply:
Of course, that makes sense, I was only thinking about the `shadow_seen.len() > 1` part, somehow I managed to skip over the other part of the condition. Could you collapse the two remaining if statements into a single one and then squash the commits which only address suggestions?